patches/squirrel.mac/feat_add_new_squirrel_mac_bundle_installation_method_behind_flag.patch

```diff
@@ -0,0 +1,128 @@
++	BOOL canRenameContentsDirectly = FALSE;
++	// For now while this is tested at scale this new option is behind a user default, this
++	// can be set by applications wishing to test this feature at runtime.  If it causes issues
++	// it can be opted out by individual users by setting this key to false explicitly.
++	// Once this has bene tested at scale it will become the default for all Squirrel.Mac
++	// users.
++	NSUserDefaults *defaults = [[NSUserDefaults alloc] init];
++	[defaults addSuiteNamed:_applicationIdentifier];
++	// In cases where this code is being executed under the ShipIt executable it's running
++	// under an application identifier equal to {parent_identifier}.ShipIt
++	// In this case we need to use the true parent identifier too as that is 99% of the time
++	// where the key will be set.
++	if ([_applicationIdentifier hasSuffix:@".ShipIt"]) {
++		[defaults addSuiteNamed:[_applicationIdentifier substringToIndex:[_applicationIdentifier length] - 7]];
++	}
++
++	if ([defaults boolForKey:@"SquirrelMacEnableDirectContentsWrite"]) {
++		canRenameContentsDirectly = [NSFileManager.defaultManager fileExistsAtPath:targetURL.path] && [NSFileManager.defaultManager fileExistsAtPath:sourceURL.path];
++
++		if (canRenameContentsDirectly) {
++			NSLog(@"Moving bundles via 'Contents' folder rename");
++		} else {
++			NSLog(@"Moving bundles directly as one of source / target does not exist.  This is unexpected.");
++		}
++	} else {
++		NSLog(@"Moving bundles directly as SquirrelMacEnableDirectContentsWrite is disabled for app: %@", _applicationIdentifier);
++	}
++	NSURL *targetContentsURL = canRenameContentsDirectly ? [targetURL URLByAppendingPathComponent:@"Contents"] : targetURL;
++	NSURL *sourceContentsURL = canRenameContentsDirectly ? [sourceURL URLByAppendingPathComponent:@"Contents"] : sourceURL;
++
+ 	return [[[[RACSignal
+ 		defer:^{
+ 			// rename() is atomic, NSFileManager sucks.
+-			if (rename(sourceURL.path.fileSystemRepresentation, targetURL.path.fileSystemRepresentation) == 0) {
++			if (rename(sourceContentsURL.path.fileSystemRepresentation, targetContentsURL.path.fileSystemRepresentation) == 0) {
+ 				return [RACSignal empty];
+ 			} else {
+ 				int code = errno;
+@@ -497,24 +538,24 @@ - (RACSignal *)installItemToURL:(NSURL *)targetURL fromURL:(NSURL *)sourceURL {
+ 			}
+ 		}]
+ 		doCompleted:^{
+-			NSLog(@"Moved bundle from %@ to %@", sourceURL, targetURL);
++			NSLog(@"Moved bundle contents from %@ to %@", sourceContentsURL, targetContentsURL);
+ 		}]
+ 		catch:^(NSError *error) {
+ 			if (![error.domain isEqual:NSPOSIXErrorDomain] || error.code != EXDEV) return [RACSignal error:error];
+ 
+ 			// If the locations lie on two different volumes, remove the
+ 			// destination by hand, then perform a move.
+-			[NSFileManager.defaultManager removeItemAtURL:targetURL error:NULL];
++			[NSFileManager.defaultManager removeItemAtURL:targetContentsURL error:NULL];
```

Comment:
Maybe out of scope for this PR, but ideally this would first move the bundle to a temp dir on the correct volume, _then_ rename.

Reply:
That's definitely a thing we can look into, but definitely also outside of this PR